source/common/formatter/substitution_formatter.h

```diff
@@ -100,10 +100,17 @@ class FormatterImpl : public Formatter {
   std::vector<FormatterProviderPtr> providers_;
 };
 
+struct JsonFormatMap;
+using JsonFormatMapValue =
+    absl::variant<const std::vector<FormatterProviderPtr>, const JsonFormatMap>;
+struct JsonFormatMap {
+  std::unique_ptr<std::map<std::string, JsonFormatMapValue>> value_;
+};
```

Comment:
Since these are only used internally, can they be made private members of JsonFormatterImpl?